### PR TITLE
Keras model to GraphDef converter

### PR DESCRIPTION
**Pull request by stephanwlee**
This change introduces Keras model serialization to GraphDef converter.
The converter will represent a model in a conceptual structure of layers
with a GraphDef.

Limitations:
- Note that this first implementation is expected to be buggy
- It can only convert Sequential and Functional models (whatever can be
    output as JSON)

### CODE REVIEW

File: tensorboard/plugins/graph/graph_util.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+          input_layer_name = scoped_name(node_name, input_layer)
+          model_inputs_to_layer[input_layer_name] = prev_node_name
+
+      # Unlike the name, there is only one item in the output_layers.
+      if layer_config.get('output_layers'):
+        maybe_layers = layer_config.get('output_layers')
+        # Normalizae output_layers.
+        # - single output: an array of values
+        # - multi outputs: an array of arrays of values
+        layers = maybe_layers if isinstance(
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]:
```suggestion
        layers = (maybe_layers if isinstance(maybe_layers[0], (list,))
                      else [maybe_layers])
```

Conforms to Google Python style and is easier to read.

File: tensorboard/plugins/graph/graph_util.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Utilities for graph plugin."""
+from tensorboard.compat.proto.graph_pb2 import GraphDef
+from tensorboard.compat.tensorflow_stub import dtypes
+
+
+def keras_model_to_graph_def(keras_layer):
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]: This function is on the longer side. Can you split up the various conditions into their own functions?

File: tensorboard/plugins/graph/graph_util.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+      node_def.attr['keras_class'].s = keras_cls_name
+
+    if layer_config.get('dtype') is not None:
+      tf_dtype = dtypes.as_dtype(layer_config.get('dtype'))
+      node_def.attr['dtype'].type = tf_dtype.as_datatype_enum
+
+    # functional configs have inbound_nodes
+    if layer.get('inbound_nodes') is not None:
+      for maybe_inbound_node in layer.get('inbound_nodes'):
+        # Normalizae inbound_nodes. It can be an array of values or an array of
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]:
```suggestion
        # Normalize inbound_nodes. It can be an array of values or an array of
```

File: tensorboard/plugins/graph/graph_util.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        # arrays of values. e.g., it can be:
+        # - [['in_layer_name', 0, 0, {}]]
+        # - [[['in_layer_is_model', 1, 0, {}], ['in_layer_is_model', 1, 1, {}]]]
+        inbound_nodes = maybe_inbound_node if isinstance(
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]:
```suggestion
        inbound_nodes = (maybe_inbound_node if isinstance(maybe_inbound_node[0], (list,))
            else [maybe_inbound_node])
```

File: tensorboard/plugins/graph/graph_util.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+  prev_node_name = None
+  for (name_scope, layer) in get_layers(keras_layer):
+    layer_config = layer.get('config')
+    node_name = scoped_name(name_scope, layer_config.get('name'))
+
+    if layer_config.get('layers'):
+      # Do not add models to GraphDef. Store information for models.
+      input_layers = layer_config.get('input_layers')
+      inbound_nodes = layer.get('inbound_nodes')
+      if input_layers and inbound_nodes:
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]: Consider adding a function that returns the type of parent and model here as an enumeration:

(parent_type, model_type) = get_parent_and_model_types(layer_config, layer)
...
if parent_type is FUNCTIONAL and model_type is FUNCTIONAL:
...
elif parent_type is FUNCTIONAL and model_type is SEQUENTIAL:
...

This will make this more readable IMO.

File: tensorboard/plugins/graph/graph_util.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+from tensorboard.compat.tensorflow_stub import dtypes
+
+
+def keras_model_to_graph_def(keras_layer):
+  """Returns GraphDef representation of the model.
+
+  Note that it only supports models that implemented get_config().
+
+  Args:
+    keras_layer: A dict from Keras model.to_json()
```
[the remark this answers is not in the document]
Comment (stephanwlee) [reply tensorboard/plugins/graph/graph_util.py]: spec is not clearly defined in a doc or anything.

File: tensorboard/plugins/graph/graph_util.py (hunk 1 above)
Comment (stephanwlee) [reply tensorboard/plugins/graph/graph_util.py]: Can you please link to the style rule? It's rather hard to read like this.

File: tensorboard/plugins/graph/graph_util_test.py [hunk 7]
```diff
@@ -0,0 +1,846 @@
...
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import json
+
+import tensorflow as tf
+
+from tensorflow.python.keras.layers import Activation, Dense, Embedding, LSTM, Input, concatenate
```
[the remark this answers is not in the document]
Comment (stephanwlee) [reply tensorboard/plugins/graph/graph_util_test.py]: While this is not a voucher to copy, here is an example: https://github.com/tensorflow/tensorboard/blob/master/tensorboard/plugins/debugger/session_debug_test.py#L38-L39

Besides extreme verbosity, I see no problem with using tf.keras.layers.*. Will make changes.

File: tensorboard/plugins/graph/graph_util.py (hunk 1 above)
Comment (manivaradarajan) [reply tensorboard/plugins/graph/graph_util.py]: https://engdoc.corp.google.com/eng/doc/devguide/py/style/index.md?cl=head#211-conditional-expressions

"Each portion must fit on one line: true-expression, if-expression, else-expression. Use a complete if statement when things get more complicated."

Recommended:

one_line = 'yes' if predicate(value) else 'no'
slightly_split = ('yes' if predicate(value)
                  else 'no, nein, nyet')

File: tensorboard/plugins/graph/graph_util.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (manivaradarajan) [reply tensorboard/plugins/graph/graph_util.py]: I thought the Cambridge folks had a schema (or at least one in progress)?

File: tensorboard/plugins/graph/graph_util.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    name_scope = keras_layer.get('config').get('name')
+    for layer in keras_layer.get('config').get('layers'):
+      for (sub_name_scope, sublayer) in _walk_layers(layer):
+        sub_name_scope = '%s/%s' % (
+            name_scope, sub_name_scope) if sub_name_scope else name_scope
+        yield (sub_name_scope, sublayer)
+
+
+def _scoped_name(name_scope, node_name):
+  """Returns scoped name for a node.
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]:
```suggestion
  """Returns scoped name for a node as a string in the form '<scope>/<node name>'.
```

File: tensorboard/plugins/graph/graph_util.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+  Returns
+    A scoped name.
+  """
+  if name_scope:
+    return '%s/%s' % (name_scope, node_name)
+  return node_name
+
+
+def _is_model(layer):
+  """Returns true if layer is a model."""
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]:
```suggestion
  """Returns True if layer is a model."""
```

File: tensorboard/plugins/graph/graph_util.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+  return layer.get('config').get('layers') is not None
+
+
+def _norm_to_list_of_layers(maybe_layers):
+  """Normalizes to a list of layers.
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]: Please mention the argument here -- "Normalizes <x> to a list of layers."

What is x?

File: tensorboard/plugins/graph/graph_util.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+    name_scope: Name scope of a model
+    model_layer: Model configuration
+    input_to_in_layer: Keras.layers.Input to inbound layer
+    model_name_to_output: Keras Model name to output layer of the model
+    prev_node_name: Previous, in sequential model layout, node name
+
+  Returns:
+    (input_to_in_layer, model_name_to_output, prev_node_name)
+
+  Two canonical types of Keras model are Functional and Sequential.
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]: Thanks this is really helpful documentation. Consider putting this at the top of the file since it's so fundamental to everything that goes on.

File: tensorboard/plugins/graph/graph_util.py [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+  return (input_to_in_layer, model_name_to_output, prev_node_name)
+
+
+def keras_model_to_graph_def(keras_layer):
+  """Returns GraphDef representation of the model.
+
+  Note that it only supports models that implemented get_config().
+
+  Args:
+    keras_layer: A dict from Keras model.to_json()
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]: Why is the argument named "keras_layer" if it's actually a model?

File: tensorboard/plugins/graph/graph_util.py [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+
+  Returns:
+    A GraphDef representation of layers.
+  """
+  input_to_layer = {}
+  model_name_to_output = {}
+  g = GraphDef()
+
+  # Sequential model layers do not have a field "inbound_nodes" but
+  # instead of defined implicitly via order of layers.
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]:
```suggestion
  # instead are defined implicitly via order of layers.
```

File: tensorboard/plugins/graph/graph_util.py [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Utilities for graph plugin."""
+from tensorboard.compat.proto.graph_pb2 import GraphDef
+from tensorboard.compat.tensorflow_stub import dtypes
+
+
+def _walk_layers(keras_layer):
+  """Walks the nested keras layer configuration in preorder.
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]: "Walks the nested keras layer configuration in preorder ... and yields <what>?"

What is the returned name_scope for?

File: tensorboard/plugins/graph/graph_util.py [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+  return (maybe_layers if isinstance(maybe_layers[0], (list,))
+          else [maybe_layers])
+
+
+def _update_dicts(name_scope,
+                  model_layer,
+                  input_to_in_layer,
+                  model_name_to_output,
+                  prev_node_name):
+  """Updates input_to_in_layer, model_name_to_output, and prev_node_name
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]: How does it update them?

File: tensorboard/plugins/graph/graph_util.py [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+      node_def.attr['keras_class'].s = keras_cls_name
+
+    if layer_config.get('dtype') is not None:
+      tf_dtype = dtypes.as_dtype(layer_config.get('dtype'))
+      node_def.attr['dtype'].type = tf_dtype.as_datatype_enum
+
+    if layer.get('inbound_nodes') is not None:
+      for maybe_inbound_node in layer.get('inbound_nodes'):
+        inbound_nodes = _norm_to_list_of_layers(maybe_inbound_node)
+        for [name, size, ind, _] in inbound_nodes:
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]: What is "ind"? I guess from reading the code it is index? Can you rename to that or something similarly descriptive?

File: tensorboard/plugins/graph/graph_util.py [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+def keras_model_to_graph_def(keras_layer):
+  """Returns GraphDef representation of the model.
+
+  Note that it only supports models that implemented get_config().
+
+  Args:
+    keras_layer: A dict from Keras model.to_json()
+
+  Returns:
+    A GraphDef representation of layers.
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]:
```suggestion
    A GraphDef representation of the layers in the model.
```

File: tensorboard/plugins/graph/graph_util.py [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+  input_layers = layer_config.get('input_layers')
+  output_layers = layer_config.get('output_layers')
+  inbound_nodes = model_layer.get('inbound_nodes')
+
+  is_functional_model = bool(input_layers and output_layers)
+  # In case of [1] and the parent model is functional, current layer
+  # will have the 'inbound_nodes' property.
+  is_parent_functional_model = bool(inbound_nodes)
+
+  if is_parent_functional_model and is_functional_model:
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]: Thanks this is much more readable.

File: tensorboard/plugins/graph/graph_util.py [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+    for (input_layer, inbound_node) in zip(input_layers, inbound_nodes):
+      input_layer_name = _scoped_name(node_name, input_layer)
+      inbound_node_name = _scoped_name(name_scope, inbound_node[0])
+      input_to_in_layer[input_layer_name] = inbound_node_name
+  elif is_parent_functional_model and not is_functional_model:
+    # Sequential model can take only one input. Make sure inbound to the
+    # model is linked to the first layer in the Sequential model.
+    prev_node_name = _scoped_name(name_scope, inbound_nodes[0][0])
+  elif not is_parent_functional_model and prev_node_name and is_functional_model:
+    assert len(input_layers) == 1, (
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/graph_util.py]: Do you want to keep this assertion here, or raise an Exception? Is this truly completely unexpected? Will the assertion completely break the graph plugin?

File: tensorboard/plugins/graph/graph_util.py (hunk 11 above)
Comment (stephanwlee) [reply tensorboard/plugins/graph/graph_util.py]: Thought about it then entire file becomes only about Keras. Would you want me to rename this module to Keras util?

File: tensorboard/plugins/graph/graph_util.py (hunk 19 above)
Comment (stephanwlee) [reply tensorboard/plugins/graph/graph_util.py]: It is truly unexpected and imo, it is incorrect to show a graphdef. I was literally thinking of breaking the application in that case.

File: tensorboard/plugins/graph/graph_util.py (hunk 15 above)
Comment (stephanwlee) [reply tensorboard/plugins/graph/graph_util.py]: How would you like me to improve name of the func and the comment? It side-effects the dicts and returns them at the end/

File: tensorboard/plugins/graph/graph_util.py (hunk 12 above)
Comment (stephanwlee) [reply tensorboard/plugins/graph/graph_util.py]: It is not a model. It is serialization of model. Since a Model is a Layer (or Network), technically speaking, a word "layer" does describe a model too.

File: tensorboard/plugins/graph/keras_util.py [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+from tensorboard.compat.tensorflow_stub import dtypes
+
+
+def _walk_layers(keras_layer):
+  """Walks the nested keras layer configuration in preorder.
+  Args:
+    keras_layer: Keras configuration from model.to_json.
+
+  Yields:
+    (name_scope, layer_config)
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/keras_util.py]: Please document the types in the returned tuple.
i.e., that name_scope is a string, layer_config is a ?

File: tensorboard/plugins/graph/keras_util.py [hunk 21]
```diff
@@ -0,0 +1,225 @@
...
+        sub_name_scope = '%s/%s' % (
+            name_scope, sub_name_scope) if sub_name_scope else name_scope
+        yield (sub_name_scope, sublayer)
+
+
+def _scoped_name(name_scope, node_name):
+  """Returns scoped name for a node as a string in the form '<scope>/<node name>'.
+
+  Args:
+    name_scope: scope names similar to that of tf.name_scope.
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/keras_util.py]:
```suggestion
    name_scope: a string representing a scope name, similar to that of tf.name_scope.
```

File: tensorboard/plugins/graph/keras_util.py [hunk 22]
```diff
@@ -0,0 +1,225 @@
...
+    node_name: current node name.
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/keras_util.py]:
```suggestion
    node_name: a string representing the current node name.
```

File: tensorboard/plugins/graph/keras_util.py [hunk 23]
```diff
@@ -0,0 +1,225 @@
...
+
+  Returns
+    A scoped name.
+  """
+  if name_scope:
+    return '%s/%s' % (name_scope, node_name)
+  return node_name
+
+
+def _is_model(layer):
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/keras_util.py]: Please document argument type.

File: tensorboard/plugins/graph/keras_util.py [hunk 24]
```diff
@@ -0,0 +1,225 @@
...
+def _update_dicts(name_scope,
+                  model_layer,
+                  input_to_in_layer,
+                  model_name_to_output,
+                  prev_node_name):
+  """Updates input_to_in_layer, model_name_to_output, and prev_node_name
+  based on the model_layer.
+
+  Args:
+    name_scope: Name scope of a model
```
Comment (manivaradarajan) [inline tensorboard/plugins/graph/keras_util.py]: Please document argument types, i.e.,

name_scope: string representing name scope of a model
input_to_in_layer: dict of Keras.layers.input to inbound layer

etc.
